### DISCUSSION TITLE
Compiling MOOSE application with static linking

### DISCUSSION THREAD

**Discussion by lekah** [idaholab/moose#22326] [Q&A Getting Started] [answered]
Dear developers and users,
We need to produce an executable for our MOOSE application that is linked static (not dynamic). I tried adapting LDFLAGS
(export LDFLAGS="-static $LDFLAGS") but it didn't produce a statically linked executable. 
Is there  a way to do this in your experience?
Many thanks in advance

**Comment by GiudGiud** [comment]
@milljm

**Reply by GiudGiud** [reply]
@permcody

**Comment by milljm** [comment, marked as answer]
Sorry for the delayed response...

You'll need to build libMesh statically as a prerequisite. MOOSE only understands how it needs to be built, based on how libMesh is built. In order to do this via our Conda packages, it will be necessary to create a new environment, which only contains what libMesh needs (moose-petsc). We also need Python. I will install moose-tools in order to obtain a version we support (3.7 -3.10):

```bash
mamba activate base    # get back to the original environment
mamba create -n petsc moose-petsc moose-tools
mamba activate petsc
```
From within this environment we can now build libMesh with static options. I am going to include some cleaning commands to clean up any previous build attempts. Keep in mind these git commands will delete any uncommitted work (so be sure to save your work if you have changes)!
```bash
cd ~/projects/moose
git clean -xfd
git submodule foreach --recursive git clean -xfd
```
Now I am going to set some environment variables to speed up my libMesh build:
```bash
export MOOSE_JOBS=28   # replace this number with how many cores your machine has
export METHODS=opt
```
We are ready to build libMesh statically:
```bash
scripts/update_and_rebuild_libmesh.sh --enable-all-static --enable-static
```
Once finished, we can now instruct MOOSE to build, and it will automatically be built statically:
```bash
cd test
make -j $MOOSE_JOBS
```

Everything we just built manually was statically included.
```bash
❯ otool -L moose_test-opt
moose_test-opt:
	@rpath/libhdf5_cpp.200.dylib (compatibility version 202.0.0, current version 202.0.0)
	@rpath/libz.1.dylib (compatibility version 1.0.0, current version 1.2.12)
	@rpath/libslepc.3.16.dylib (compatibility version 3.16.0, current version 3.16.2)
	@rpath/libpetsc.3.16.dylib (compatibility version 3.16.0, current version 3.16.6)
	@rpath/libHYPRE-2.23.0.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libstrumpack.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libsuperlu_dist.7.dylib (compatibility version 7.0.0, current version 7.1.1)
	@rpath/libhdf5_hl.200.dylib (compatibility version 201.0.0, current version 201.1.0)
	@rpath/libhdf5.200.dylib (compatibility version 202.0.0, current version 202.0.0)
	@rpath/libparmetis.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libmetis.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libmpifort.12.dylib (compatibility version 15.0.0, current version 15.2.0)
	@rpath/libpmpi.12.dylib (compatibility version 15.0.0, current version 15.2.0)
	@rpath/libgfortran.5.dylib (compatibility version 6.0.0, current version 6.0.0)
	@rpath/libquadmath.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.100.3)
	@rpath/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libgcc_s.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libmpi.12.dylib (compatibility version 15.0.0, current version 15.2.0)
	@rpath/libmpicxx.12.dylib (compatibility version 15.0.0, current version 15.2.0)
	@rpath/libomp.dylib (compatibility version 5.0.0, current version 5.0.0)
```
There should be no mention of libraries which we built statically.

**Reply by roystgnr** [reply]
You might need to specify `--enable-static` too?  I've never used static linking myself, and I'm not sure how we propagate the options down to dependencies, but it looks like `--enable-all-static` is a libMesh-specific thing that TIMPI might be ignoring, whereas `--enable-static` is a libtool standard option.

**Reply by milljm** [reply]
That did the trick. Thanks @roystgnr 

I edited my original response to include the additional switch.

**Reply by lekah** [reply]
Thank you very much @milljm @roystgnr
